### PR TITLE
drivers: usb: Fix NXP driver for CDC test cases

### LINKED ISSUE #52625: USB device: mimxrt685_evk_cm33: premature ZLP during control IN transfer
**Describe the bug**
USB CDC ACM samples does not work on NXP MIMXRT685-EVK (mimxrt685_evk_cm33).

The behavior is the same for both v3.2.0 and current main branch. The device prematurely ends GET CONFIGURATION control transfer. During Data stage, the first 64 bytes are transferred correctly, then instead of sending remaining 11 bytes, the device sends ZLP that ends the Control transfer Data stage.

**To Reproduce**
1. west build --pristine -b mimxrt685_evk_cm33 -d build-cdc-nxp685 ~/zephyrproject/zephyr/samples/subsys/usb/cdc_acm
2. cd build-cdc-nxp685
3. west flash
4. Connect RT685 USB to host and observe enumeration failure

**Expected behavior**
Host should receive whole CONFIGURATION descriptor, i.e. all requested 75 bytes 

**Impact**
This renders USB device support useless for all classes that use control IN transfers with wLength larger than 64 bytes.

**Logs and console output**
```
[15687.689485] usb 3-4: new high-speed USB device number 32 using xhci_hcd
[15687.836726] usb 3-4: config index 0 descriptor too short (expected 75, got 64)
[15687.836741] usb 3-4: config 1 has an invalid descriptor of length 7, skipping remainder of the config
[15687.836749] usb 3-4: config 1 interface 1 altsetting 0 has 0 endpoint descriptors, different from the interface descriptor's value: 2
[15687.837223] usb 3-4: New USB device found, idVendor=2fe3, idProduct=0001, bcdDevice= 3.02
[15687.837234] usb 3-4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[15687.837239] usb 3-4: Product: Zephyr CDC ACM sample
[15687.837243] usb 3-4: Manufacturer: ZEPHYR
[15687.837246] usb 3-4: SerialNumber: 6A631C9D80D5EC67
[15687.839685] cdc_acm: probe of 3-4:1.0 failed with error -22
```

USB connection trace can be viewed in Wireshark 4.0 or newer.
[zephyr-cdc-acm-nxp-mimxrt685_evk_cm33.pcap.gz](https://github.com/zephyrproject-rtos/zephyr/files/10112751/zephyr-cdc-acm-nxp-mimxrt685_evk_cm33.pcap.gz)

Packet 816 in attached capture contains the ZLP that prematurely ends Data stage. Instead of the ZLP, the device should have sent the trailing CONFIGURATION descriptor 11 bytes.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK 0.15.1
 - zephyr-v3.2.0 tag (4256cd41df6c60f1832fd2deb14edc30ac7debab)

**Additional context**
The issue seems to be in the NXP driver and not in USB stack because the same sample works perfectly fine on Nordic targets, e.g. nrf52840dk_nrf52840

### LINKED ISSUE #53636: cdc_acm fails on lpcxpresso55s69 board
**Describe the bug**
Running the cdc_acm sample on this board fails.  

**To Reproduce**
Tested with this build command
```
west build -b lpcxpresso55s69_cpu0 samples/subsys/usb/cdc_acm
```

**Logs and console output**
With the latest main branch, the USB device does not enumerate properly in Windows.  The debug console shows no error:
```
*** Booting Zephyr OS build zephyr-v3.2.0-3235-g871d6d26eeb0 ***
[00:00:00.000,000] <inf> cdc_acm_echo: Wait for DTR
```

Using v3.2.0, the USB device enumerates, but gives this error:
```
*** Booting Zephyr OS build zephyr-v3.2.0  ***
[00:00:00.004,000] <inf> cdc_acm_echo: Wait for DTR
[00:00:07.560,000] <err> usb_dc_mcux: Failed to fill ep 0x80 buffer
```

Testing in v3.1, the app crashes with a hard fault:
```
*** Booting Zephyr OS build zephyr-v3.1.0  ***
[00:00:00.004,000] <err> os: ***** HARD FAULT *****
[00:00:00.004,000] <err> os:   Bus fault on vector table read
[00:00:00.004,000] <err> os: r0/a1:  0xfefa125b  r1/a2:  0x2280ff63  r2/a3:  0x4
0000000
[00:00:00.004,000] <err> os: r3/a4:  0x300007b0 r12/ip:  0x300007b0 r14/lr:  0x3
00039b0
[00:00:00.004,000] <err> os:  xpsr:  0xc33cc200
[00:00:00.004,000] <err> os: Faulting instruction address (r15/pc): 0x5ac3c35a
[00:00:00.004,000] <err> os: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
[00:00:00.004,000] <err> os: Current thread: 0x30000cc8 (unknown)
[00:00:00.061,000] <err> os: Halting system
```

**Environment (please complete the following information):**
 - tested on Windows10
 - Toolchain Zephyr SDK 0.15.1

### PR DESCRIPTION
**Pull request by mmahadevan108**
The cdc_acm tests were failing. Remove a length check in the write function as this is handled by the HAL driver. Remove unused type member and improve the error checking.

Fixes #53636 , #52625

### CODE REVIEW

**Review by desowin — DISMISSED** [review]

**Review by jfischer-no — CHANGES_REQUESTED** [review]
Please split it in three commits, one for cleanup and error checking, second for the changes in usb_dc_ep_read_wait, and another for the changes in usb_dc_ep_write() to fix control IN transfers on IP3511.

File: drivers/usb/device/usb_dc_mcux.c [hunk 1]
```diff
@@ -490,37 +487,35 @@ int usb_dc_ep_write(const uint8_t ep, const uint8_t *const data,
...
-		len_to_send = dev_state.eps[ep_abs_idx].ep_mps;
-	} else {
-		len_to_send = data_len;
+	if (USB_EP_GET_DIR(ep) != USB_EP_DIR_IN) {
+		LOG_ERR("Wrong endpoint direction");
+		return -EINVAL;
 	}
 
-	for (uint32_t n = 0; n < len_to_send; n++) {
+	for (uint32_t n = 0; n < data_len; n++) {
```
[the remark this answers is not in the document]
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c]: I have removed this copy.

**Comment by mmahadevan108** [comment]
> Please split it in three commits, one for cleanup and error checking, second for the changes in usb_dc_ep_read_wait, and another for the changes in usb_dc_ep_write() to fix control IN transfers on IP3511.

PR split into 3 commits. I have tested this PR for both the EHCI (RT1060) and IP3511 (LPC55S69) controllers

**Review by desowin — DISMISSED** [review]

File: drivers/usb/device/usb_dc_mcux.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jfischer-no) [reply drivers/usb/device/usb_dc_mcux.c]: Okay, but please check if there are requirements regarding alignment and memory type for IP3511 and EHCI controller, data can also point to an address in flash.

[the remark this answers is not in the document]
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c]: I have not seen any limitations. I have tested using various Zephyr USB samples such as `cdc_acm`, `mass`, `hid `etc.

File: drivers/usb/device/usb_dc_mcux.c [hunk 2]
```diff
@@ -486,8 +486,6 @@ int usb_dc_ep_write(const uint8_t ep, const uint8_t *const data,
 		    const uint32_t data_len, uint32_t *const ret_bytes)
 {
 	uint8_t ep_abs_idx = EP_ABS_IDX(ep);
-	uint8_t *buffer = (uint8_t *)dev_state.eps[ep_abs_idx].block.data;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_mcux.c:492]: This intermediate buffer is not needed anymore. 
https://github.com/zephyrproject-rtos/zephyr/blob/e7443bb2035574ec1377b791032769e3d0a7f428/drivers/usb/device/usb_dc_mcux.c#L281
Should it still be allocated?

Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c:492]: This is used for the receive.

Comment (jfischer-no) [reply drivers/usb/device/usb_dc_mcux.c:492]: It is allocated for both IN and OUT endpoints, but with this changes the driver does not need to allocate it for the IN endpoints.

**Comment by mmahadevan108** [comment]
@jfischer-no @desowin , can you take a look again.

**Review by jfischer-no — APPROVED** [review]
Maybe you want to reduce number of blocks as well,
`#define EP_BUF_NUMOF_BLOCKS	((NUM_OF_EP_MAX + 3) / 4)`

Otherwise LGTM, although I have not tested due to lack of time.

**Review by tmon-nordic — APPROVED** [review]

**Pull request merged**